### PR TITLE
quay.io pull through cache image

### PR DESCRIPTION
**Pull request by jmguzik**
Signed-off-by: Jakub Guzik <jguzik@redhat.com>

### CODE REVIEW

File: images/quayio-pull-through-cache/docker-entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+set -e
+
+secret_path=/etc/pull-secret/.dockerconfigjson
+ls -l -S $secret_path 2>/dev/null || echo "Missing $secret_path" && exit 1
+export REGISTRY_PROXY_USERNAME=$(cat $secret_path | jq '.auths["quay.io"].email' | cut -d '"' -f 2)
```
Comment (jupierce) [inline images/quayio-pull-through-cache/docker-entrypoint.sh:6]: The the case of our service account, I think username differs from email. Username looks like `openshift-release-dev+ocm_access_7bf7....` and email looks like `aos-team-dp-testplatform@redhat.com` in the secret. Docker buries username in the `auth` field.
`jq '.auths["quay.io"].auth' -r | base64 -d`  will show you the username/password pair. They are colon delimited.

File: images/quayio-pull-through-cache/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	case "$arch" in \
+		x86_64)  arch='amd64';; \
+		aarch64) arch='arm64';; \
+		armhf)   arch='armv6';; \
+		armv7)   arch='armv7';; \
+		ppc64le) arch='ppc64le';; \
+		s390x)   arch='s390x';; \
+		*) echo >&2 "error: unsupported architecture: $arch"; exit 1 ;; \
+	esac; \
+	wget -O registry.tar.gz "https://github.com/distribution/distribution/releases/download/v${version}/registry_${version}_linux_${arch}.tar.gz"; \
```
Comment (jupierce) [inline images/quayio-pull-through-cache/Dockerfile:18]: We should exercise caution here. Major changes could take place in the implementation. Config options, server API, changes that require data migration / etc (e.g. when 3.0 comes up). Given the impacts of the service failing to start, I'd lean toward this being an intentional bump or at least constraining it to a major/minor.

File: images/quayio-pull-through-cache/docker-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+export REGISTRY_PROXY_PASSWORD=$(cat $secret_path | jq '.auths["quay.io"].auth' | cut -d '"' -f 2)
+htpasswd -Bbn ${REGISTRY_PROXY_USERNAME} ${REGISTRY_PROXY_PASSWORD} > /etc/htpasswd
+
+config="/etc/docker/registry/config.yml"
+ls -l -S $config 2>/dev/null || echo "Missing $config" && exit 1
+registry serve $config &
+pid=$!
+
+while true; do
+    quay_io_username=$(cat $secret_path | jq '.auths["quay.io"].email' | cut -d '"' -f 2)
```
Comment (jupierce) [inline images/quayio-pull-through-cache/docker-entrypoint.sh:16]: Same note on extracting the username / password

File: images/quayio-pull-through-cache/Dockerfile (hunk 2 above)
Comment (jmguzik) [reply images/quayio-pull-through-cache/Dockerfile:18]: Docker hardcoded this to 2.8.1 and does a manual upgrade. We can do it similarly.

Comment (jmguzik) [reply images/quayio-pull-through-cache/Dockerfile:16]: Fixed to 2.8.1, updates are very rare anyway

File: images/quayio-pull-through-cache/docker-entrypoint.sh (hunk 3 above)
Comment (jmguzik) [reply images/quayio-pull-through-cache/docker-entrypoint.sh]: fixed

**Pull request merged**